File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        

        Stack<Integer> st = new Stack<>();
        int[] nge = new int[nums2.length];

        Map<Integer, Integer> mp = new HashMap<>();

        for (int i = 0; i< nums2.length;i++){
            mp.put(nums2[i], i);
        }


        for (int i = nums2.length-1; i>=0;i--){
            
            while(!st.isEmpty() && nums2[i] > st.peek()){
                st.pop();
            }
            if (st.isEmpty()) nge[i] = -1;
            else nge[i] = st.peek();

            st.push(nums2[i]);
        }

        int[] ans = new int[nums1.length];

        for (int i = 0; i< nums1.length;i++){
            int cur = nums1[i];
            ans[i] = nge[mp.get(cur)];

        }

        return ans;
        
    }
}

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st = []
        ans = []
        nge = [-1]*len(nums2)
        for i in range(len(nums2)-1, -1, -1):
            num = nums2[i]
            while st and num >= st[-1]:
                st.pop()

            if st:
                nge[i] = st[-1]

            st.append(num)

        mp = {}
        for num in nums1:
            mp[num] = -1

        for i,num in enumerate(nums2):
            if num in mp:
                mp[num] = i

        for k,v in mp.items():
            if mp[k] != -1:
                ans.append(nge[mp[k]])
            else:
                ans.append(-1)

        return ans
